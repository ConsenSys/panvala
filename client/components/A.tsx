import styled from 'styled-components';
import { color, typography } from 'styled-system';

const A: any = styled.a`
  color: inherit;
  cursor: pointer;
  ${color};
  ${({ bold }) => bold && 'font-weight: bold'};
`;

export default A;
