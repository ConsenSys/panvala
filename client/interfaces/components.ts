import * as React from 'react';

export interface StatelessPage<P = {}> extends React.SFC<P> {
  getInitialProps?: (ctx: any) => Promise<P>;
}

export interface IParameterChangesObject {
  [key: string]: {
    oldValue: string;
    newValue: string;
  };
}

export interface IGovernanceSlateFormValues {
  email: string;
  title: string;
  firstName: string;
  lastName?: string;
  organization?: string;
  summary: string;
  recommendation: string;
  parameters: IParameterChangesObject;
  stake: string;
}
