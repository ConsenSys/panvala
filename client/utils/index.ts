// Re-export Panvala utils to be used throughout
export { default as panvala_utils } from '../../packages/panvala-utils';
