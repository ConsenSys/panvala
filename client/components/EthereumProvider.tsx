import * as React from 'react';
import { utils, providers } from 'ethers';
import { toast } from 'react-toastify';
import isEmpty from 'lodash/isEmpty';
import getConfig from 'next/config';
import { connectProvider, connectContracts } from '../utils/provider';
import { IContracts } from '../interfaces';
import { baseToConvertedUnits, BN } from '../utils/format';
import {
  saveState,
  loadState,
  ENABLED_ACCOUNTS,
  saveSessionState,
  CLOSED_MAINNET_MODAL,
  loadSessionState,
} from '../utils/localStorage';
import MainnetModal from './MainnetModal';
const { publicRuntimeConfig = {} } = getConfig() || {};

export interface IEthereumContext {
  account: string;
  ethProvider: providers.Web3Provider;
  contracts: IContracts;
  panBalance: utils.BigNumber;
  gkAllowance: utils.BigNumber;
  tcAllowance: utils.BigNumber;
  votingRights: utils.BigNumber;
  slateStakeAmount: utils.BigNumber;
  onRefreshBalances(): void;
}
export const EthereumContext: React.Context<IEthereumContext> = React.createContext<any>({});

// extend the global object to include ethereum (metamask ethereum provider)
declare global {
  interface Window {
    ethereum: any;
  }
}

function reducer(state: any, action: any) {
  switch (action.type) {
    case 'chain':
      return {
        ...state,
        ethProvider: action.ethProvider,
        account: action.account,
      };
    case 'contracts':
      return {
        ...state,
        contracts: action.contracts,
        slateStakeAmount: action.slateStakeAmount,
      };
    case 'balances':
      return {
        ...state,
        ...action.payload,
      };
    default: {
      console.error('Unknown action type provided to Eth Provider', action);
      return state;
    }
  }
}

const EthereumProvider: React.FC<any> = (props: any) => {
  const [state, dispatch] = React.useReducer(reducer, {
    ethProvider: {},
    contracts: {
      gatekeeper: {},
      tokenCapacitor: {},
      token: {},
      parameterStore: {},
    },
    account: '',
    panBalance: BN('0'),
    gkAllowance: BN('0'),
    tcAllowance: BN('0'),
    votingRights: BN('0'),
    tcBalance: BN('0'),
    gkBalance: BN('0'),
    slateStakeAmount: BN('0'),
  });

  // runs once, on-load
  // 1. get metamask / set account / set provider
  // 2. set contracts
  React.useEffect(() => {
    async function handleConnectEthereum() {
      try {
        if (typeof window !== 'undefined' && window.hasOwnProperty('ethereum')) {
          // this means metamask is installed. get the ethereum provider
          const { ethereum }: Window = window;
          // pop-up metamask to authorize panvala-app (account signature validation)
          let addresses: string[] = [];
          try {
            addresses = await ethereum.enable();
            let enabledAccounts = loadState(ENABLED_ACCOUNTS);
            if (enabledAccounts && !enabledAccounts.includes(addresses[0])) {
              enabledAccounts = enabledAccounts.concat(addresses[0]);
            } else if (!enabledAccounts) {
              enabledAccounts = addresses;
            }
            saveState(ENABLED_ACCOUNTS, enabledAccounts);
          } catch (error) {
            console.error(`ERROR failed to enable metamask: ${error.message}`);
            throw error;
          }

          // selected account
          const account = utils.getAddress(addresses[0]);

          // register an event listener to handle account-switching in metamask
          ethereum.once('accountsChanged', (accounts: string[]) => {
            console.info('MetaMask account changed:', accounts[0]);
            handleConnectEthereum();
          });

          // register an event listener to handle network-switching in metamask
          ethereum.once('networkChanged', (network: any) => {
            console.info('MetaMask network changed:', network);
            window.location.reload();
          });

          try {
            // wrap MetaMask with ethers
            const ethProvider = await connectProvider(ethereum);
            console.log('account:', account);

            dispatch({
              type: 'chain',
              ethProvider,
              account,
            });

            if (!isEmpty(ethProvider)) {
              await setupContracts(ethProvider);
            }
          } catch (error) {
            console.error(`ERROR failed to connect eth provider: ${error.message}`);
            throw error;
          }

          async function setupContracts(ethProvider) {
            // contract abstractions (w/ metamask signer)
            try {
              const contracts = await connectContracts(ethProvider);
              const slateStakeAmount = await contracts.parameterStore.functions.getAsUint(
                'slateStakeAmount'
              );
              console.log('contracts:', contracts);

              dispatch({
                type: 'contracts',
                contracts,
                slateStakeAmount,
              });
            } catch (error) {
              console.error(`ERROR failed to connect contracts: ${error.message}`);
              if (error.message.includes('contract not deployed')) {
                toast.error(`Contracts not deployed on current network`);
              } else if (error.message.includes('Wrong network')) {
                toast.error(error.message);
              }
              throw error;
            }
          }
        }
      } catch (error) {
        console.error(error);
      }
    }
    handleConnectEthereum();

    // clean up
    return () =>
      typeof window !== 'undefined' &&
      typeof window.ethereum !== 'undefined' &&
      window.ethereum.removeAllListeners();
  }, []);

  async function handleRefreshBalances() {
    const {
      account,
      contracts: { token, gatekeeper, tokenCapacitor },
    } = state;
    if (account && !isEmpty(token) && !isEmpty(gatekeeper)) {
      // prettier-ignore
      const [panBalance, gkAllowance, tcAllowance, votingRights, tcBalance, gkBalance] = await Promise.all([
        token.functions.balanceOf(account),
        token.functions.allowance(account, gatekeeper.address),
        token.functions.allowance(account, tokenCapacitor.address),
        gatekeeper.functions.voteTokenBalance(account),
        token.functions.balanceOf(tokenCapacitor.address),
        token.functions.balanceOf(gatekeeper.address),
      ]);
      const balances = {
        panBalance,
        gkBalance,
        tcBalance,
        gkAllowance,
        tcAllowance,
        votingRights,
      };
      const converted = {
        votingRights: baseToConvertedUnits(votingRights),
        balance: baseToConvertedUnits(panBalance),
        gatekeeper: baseToConvertedUnits(gkBalance),
        tokenCapacitor: baseToConvertedUnits(tcBalance),
        gkAllowance: baseToConvertedUnits(gkAllowance),
        tcAllowance: baseToConvertedUnits(tcAllowance),
      };
      console.log('balances:', converted);
      dispatch({
        type: 'balances',
        payload: balances,
      });
    }
  }

  // runs whenever account changes
  React.useEffect(() => {
    if (state.account && !isEmpty(state.contracts.token)) {
      console.info('Refreshing balances for:', state.account.slice(0, 10));
      handleRefreshBalances();
    }
  }, [state.account, state.contracts.token]);

  // render warning modal if on mainnet
  const [modalIsOpen, setMainnetModalOpen] = React.useState(false);
  React.useEffect(() => {
    async function checkNetwork() {
      const network = await state.ethProvider.getNetwork();
      if (network.chainId === 1 || publicRuntimeConfig.panvalaEnv === 'production') {
        const sessionState = loadSessionState(CLOSED_MAINNET_MODAL);
        if (sessionState !== 'TRUE') {
          setMainnetModalOpen(true);
        }
      }
    }

    if (!isEmpty(state.ethProvider)) {
      checkNetwork();
    }
  }, [state.ethProvider]);

  const ethContext: IEthereumContext = {
    ...state,
    onRefreshBalances: handleRefreshBalances,
  };

  function closeModal() {
    saveSessionState(CLOSED_MAINNET_MODAL, 'TRUE');
    setMainnetModalOpen(false);
  }

  return (
    <EthereumContext.Provider value={ethContext}>
      {props.children}

      <MainnetModal modalIsOpen={modalIsOpen} handleClick={closeModal} />
    </EthereumContext.Provider>
  );
};

export default React.memo(EthereumProvider);
