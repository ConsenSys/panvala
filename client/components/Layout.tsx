import * as React from 'react';
import Head from 'next/head';
import styled from 'styled-components';
import { ToastContainer, toast } from 'react-toastify';
import { format } from 'date-fns';
import { AxiosResponse } from 'axios';

import Header from './Header';
import { IProposal, ISlate, IAppContext } from '../interfaces';
import { getAllProposals, getAllSlates } from '../utils/api';
import { baseToConvertedUnits } from '../utils/format';

export const AppContext: React.Context<IAppContext> = React.createContext<IAppContext>({
  slates: [],
  proposals: [],
  currentBallot: {
    startDate: 0,
    votingOpenDate: 0,
    votingCloseDate: 0,
    finalityDate: 0,
  },
  notifications: [],
});

export const LayoutWrapper = styled.div`
  font-family: 'Roboto';
  min-height: 100vh;
  margin: 2em 10em;
`;

type IProps = {
  title: string;
  children: any;
};

export default class Layout extends React.Component<IProps, IAppContext> {
  readonly state: IAppContext = {
    slates: [],
    proposals: [],
    currentBallot: {
      startDate: 0,
      votingOpenDate: 0,
      votingCloseDate: 0,
      finalityDate: 0,
    },
    notifications: [{ action: 'Welcome to Panvala!' }],
    onNotify: () => this.handleNotify,
    onRefreshProposals: () => this.handleRefreshProposals(),
    onRefreshSlates: () => this.handleRefreshSlates(),
    onHandleNotification: (n: any) => this.handleNotification(n),
  };

  constructor(props: IProps) {
    super(props);
    this.handleNotify = this.handleNotify.bind(this);
    this.handleRefreshProposals = this.handleRefreshProposals.bind(this);
    this.handleRefreshSlates = this.handleRefreshSlates.bind(this);
    this.handleNotification = this.handleNotification.bind(this);
  }

  // runs once, onload
  async componentDidMount() {
    // const slatesFromIpfs: any[] = await Promise.all(slateMultihashes.map(mh => ipfsGetData(mh)));
    // const slatesWithMHs = slates.map((s, i) => ({ ...s, hash: slateMultihashes[i] }));
    let slates: ISlate[] = await this.handleGetAllSlates();
    const proposals: IProposal[] | AxiosResponse = await this.handleGetAllProposals();

    let proposalData: IProposal[] = [];
    // sort proposals by createdAt
    if (Array.isArray(proposals)) {
      const sortedProposals = proposals.sort((a: IProposal, b: IProposal) => {
        const timestampA = format(a.createdAt, 'x');
        const timestampB = format(b.createdAt, 'x');
        return parseInt(timestampA) - parseInt(timestampB);
      });
      proposalData = sortedProposals;
    }

    const oneWeekSeconds = 604800;
    // Epoch 3
    // beginning of week 1 (2/1)
    const epochStartDate = 1549040401; // gateKeeper.functions.currentBatchStart()
    // end of week 11 (4/19)
    const week11EndDate = epochStartDate + oneWeekSeconds * 11; // 1555689601
    // end of week 12 (4/26)
    const week12EndDate = week11EndDate + oneWeekSeconds;
    // end of week 13 (5/3)
    const week13EndDate = week12EndDate + oneWeekSeconds;

    // prettier-ignore
    this.setState({
      slates,
      proposals: proposalData,
      currentBallot: {
        startDate: epochStartDate,
        votingOpenDate: week11EndDate,
        votingCloseDate: week12EndDate,
        finalityDate: week13EndDate,
      },
    });
  }

  handleNotify(note: string, custom: string) {
    if (custom) {
      (toast as any)[custom](note);
    } else {
      toast(note);
    }
  }

  async handleGetAllProposals() {
    let proposals: IProposal[] | AxiosResponse = await getAllProposals();

    // convert tokensRequested: base -> human
    if (Array.isArray(proposals)) {
      proposals = proposals.map((p: any) => {
        p.tokensRequested = baseToConvertedUnits(p.tokensRequested, 18);
        return p;
      });
    } else {
      throw new Error('invalid proposals type -- AxiosResponse');
    }

    return proposals;
  }

  async handleRefreshProposals() {
    const proposals: IProposal[] = await this.handleGetAllProposals();
    return this.setState({ proposals });
  }

  async handleGetAllSlates() {
    const slates: ISlate[] | AxiosResponse = await getAllSlates();
    console.log('slates:', slates);
    if (Array.isArray(slates)) {
      return slates;
    }

    return [];
  }

  async handleRefreshSlates() {
    const slates: ISlate[] = await this.handleGetAllSlates();
    return this.setState({ slates });
  }

  handleNotification(notification: any) {
    this.setState({
      notifications: [...this.state.notifications, notification],
    });
  }

  render() {
    const { children, title }: IProps = this.props;
    console.log('Layout state:', this.state);

    return (
      <LayoutWrapper>
        <Head>
          <title>{title}</title>
        </Head>

        <Header notifications={this.state.notifications} />
        <AppContext.Provider value={this.state}>{children}</AppContext.Provider>
        <ToastContainer
          position="bottom-right"
          autoClose={8000}
          hideProgressBar={true}
          newestOnTop={false}
          rtl={false}
          draggable={false}
          closeOnClick
          pauseOnHover
        />
      </LayoutWrapper>
    );
  }
}
