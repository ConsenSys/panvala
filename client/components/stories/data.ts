import Router from 'next/router';
import range from 'lodash/range';
import uuid from 'uuid/v4';
import { utils } from 'ethers';
import { ballotDates } from '../../utils/voting';
import { ISlate, IProposal } from '../../interfaces';
import '../../globalStyles.css';

// https://github.com/zeit/next.js/issues/1827#issuecomment-323314141
(Router as any).router = {
  push: () => {},
};

export const epochStartDate = 1541178000; // Nov 2, 2018 17:00:00 GMT
export const currentBallot = ballotDates(epochStartDate);
export const baseBallot = {
  epochNumber: 0,
  startDate: 1541178000,
  initialSlateSubmissionDeadline: 1544504400,
  votingOpenDate: 1547830800,
  votingCloseDate: 1548435600,
  finalityDate: 1549040400,
  slateSubmissionDeadline: { GRANT: 0, GOVERNANCE: 0 },
};

export const proposals: IProposal[] = [
  {
    id: 0,
    category: 'GRANT',
    firstName: 'Amanda',
    lastName: 'Crypto',
    title: 'A great project',
    tokensRequested: '5000000000',
    summary:
      "Plasma is a family of protocols which allow individuals to easily deploy high-throughput, secure blockchains. A smart contract on Ethereum’s main chain can ensure that users’ funds are secure, even if the “plasma chain” acts fully maliciously. This eliminates the need for a trusted pegging mechanism like that of sidechains. Plasma chains are non-custodial, allowing the prioritization of scalability without sacrificing security. We’ve devised a new architecture for building Plasma apps on one generalized plasma chain. It establishes a clean separation between the plasma layer and the application layer. We will publish a generalized Plasma predicate contract framework, which allows for upgradeability and composability of plasma contracts. Since plasma research moves so quickly, we realized we needed to develop an architecture that allowed for maximal modularity, to prevent vast chunks of code from being thrown away with each new research discovery. With that framework, we will refactor our existing codebase for secure payments using predicates. We want to use the generalized Plasma research we did and put it to the test in our codebase refactor following the launch of our testnet on January 31. Dogfooding is a critical part of the work we do, as there's no point to open sourcing the codebase if it's not readable or easy to use.",
    awardAddress: '0xd115bffabbdd893a6f7cea402e7338643ced44a6',
    projectTimeline: 'timeline',
    teamBackgrounds: 'backgrounds',
  },
];

export const unstakedSlate: ISlate = {
  id: 0,
  category: 'GRANT',
  status: 0,
  owner: 'John Doe',
  recommender: '0xd115bffabbdd893a6f7cea402e7338643ced44a6',
  organization: 'Team Recommender',
  description:
    'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer eu nibh molestie, auctor ligula a, faucibus ante. Morbi dapibus enim in vulputate congue. Mauris feugiat gravida nibh, sed pellentesque eros pellentesque eu. Sed rutrum vitae magna sed aliquet. Suspendisse facilisis vulputate lobortis. Vestibulum sed dolor eu mi molestie pharetra. Duis ut diam aliquam, molestie erat non, scelerisque ligula. Curabitur accumsan ipsum pellentesque posuere ornare. Sed vulputate cursus accumsan. Morbi efficitur dictum magna, a imperdiet mauris aliquet vitae.',
  proposals,
  requiredStake: utils.bigNumberify('300000000000000000000'),
  verifiedRecommender: false,
  incumbent: false,
  staker: '0x0000000000000000000000000000000000000000',
  epochNumber: 1,
};

export function makeBallot(options?: any, prototype?: any) {
  return {
    ...baseBallot,
    ...prototype,
    ...options,
  };
}

export function makeSlate(options: any, prototype?: ISlate) {
  return {
    ...unstakedSlate,
    ...prototype,
    ...options,
    id: uuid(),
  };
}

export function makeSlates(num: number, prototype: ISlate) {
  return range(1, num).map(() => makeSlate({}, prototype));
}

export const acceptedGrantSlate = makeSlate({
  verifiedRecommender: true,
  status: 3,
  epochNumber: 0,
});

export const rejectedGrantSlate = makeSlate(
  {
    status: 2,
  },
  acceptedGrantSlate
);

export const acceptedGovSlate = makeSlate(
  {
    category: 'GOVERNANCE',
    verifiedRecommender: true,
    status: 3,
  },
  acceptedGrantSlate
);

export const rejectedGovSlate = makeSlate(
  {
    status: 2,
  },
  acceptedGovSlate
);
