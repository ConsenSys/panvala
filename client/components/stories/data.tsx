import Router from 'next/router';
import { utils } from 'ethers';
import { ballotDates } from '../../utils/status';
import { ISlate, IProposal } from '../../interfaces';
import '../../globalStyles.css';

// https://github.com/zeit/next.js/issues/1827#issuecomment-323314141
(Router as any).router = {
  push: () => {},
  prefetch: () => {},
};

export const epochStartDate = 1549040401;
export const currentBallot = ballotDates(15499990);

export const proposals: IProposal[] = [
  {
    id: 0,
    firstName: 'Amanda',
    lastName: 'Crypto',
    title: 'A great project',
    tokensRequested: '5000000000',
    summary: 'All sorts of amazing things',
    awardAddress: '0xd115bffabbdd893a6f7cea402e7338643ced44a6',
    projectTimeline: 'timeline',
    teamBackgrounds: 'backgrounds',
  },
];

export const unstakedSlate: ISlate = {
  id: 0,
  category: 'GRANT',
  status: 0,
  deadline: currentBallot.votingOpenDate,
  title: 'Some slate',
  owner: 'John Doe',
  recommenderAddress: '0xd115bffabbdd893a6f7cea402e7338643ced44a6',
  organization: 'Team Recommender',
  description:
    'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer eu nibh molestie, auctor ligula a, faucibus ante. Morbi dapibus enim in vulputate congue. Mauris feugiat gravida nibh, sed pellentesque eros pellentesque eu. Sed rutrum vitae magna sed aliquet. Suspendisse facilisis vulputate lobortis. Vestibulum sed dolor eu mi molestie pharetra. Duis ut diam aliquam, molestie erat non, scelerisque ligula. Curabitur accumsan ipsum pellentesque posuere ornare. Sed vulputate cursus accumsan. Morbi efficitur dictum magna, a imperdiet mauris aliquet vitae.',
  proposals,
  requiredStake: utils.bigNumberify('300000000000000000000'),
  verifiedRecommender: false,
};
