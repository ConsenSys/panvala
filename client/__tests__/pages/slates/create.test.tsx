import * as React from 'react';
import { render } from 'react-testing-library';
import 'jest-dom/extend-expect';

import CreateSlate from '../../../pages/slates/create';
import { AppContext } from '../../../components/Layout';
import { proposalsArray } from '../../../utils/data';
import { EthereumContext } from '../../../components/EthereumProvider';

describe('CreateSlate', () => {
  it('should render correctly pages/slates/create form component', () => {
    const { container, getByText } = render(
      <AppContext.Provider value={{ proposals: proposalsArray }}>
        <EthereumContext.Provider value={{ contracts: {} }}>
          <CreateSlate />
        </EthereumContext.Provider>
      </AppContext.Provider>
    );
    expect(container).toMatchSnapshot();
    expect(getByText('Select the grants that you would like to add to your slate')).toHaveClass(
      'mv3 f7 black-50'
    );
  });
});
