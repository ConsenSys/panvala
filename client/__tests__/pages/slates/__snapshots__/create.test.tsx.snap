// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateSlate should render correctly pages/slates/create form component 1`] = `
<div>
  <div>
    <div
      class="b f2 tc mv4"
    >
      Create a Grant Slate
    </div>
    <div
      class="Form__FormWrapper-naepsk-1 ciGIOt"
    >
      <form>
        <div
          class="pa4"
        >
          <div
            class="SectionLabel__StyledSectionLabel-sc-1fzxsyz-0 hMXPaX"
          >
            ABOUT
          </div>
          <label
            class="Label__StyledLabel-sc-1pdmyea-0 fEfxuf"
            for="email"
            required=""
          >
            Email
            <span
              class="Required__StyledRequired-sc-1du3aep-0 icgBVx"
            >
               *
            </span>
          </label>
          <input
            class="FieldText__Field-sc-1uqpfc5-0 dhVsEX"
            name="email"
            placeholder="Enter your email"
            type="text"
            value=""
          />
          <label
            class="Label__StyledLabel-sc-1pdmyea-0 fEfxuf"
            for="title"
            required=""
          >
            Title
            <span
              class="Required__StyledRequired-sc-1du3aep-0 icgBVx"
            >
               *
            </span>
          </label>
          <input
            class="FieldText__Field-sc-1uqpfc5-0 dhVsEX"
            name="title"
            placeholder="Enter title"
            type="text"
            value=""
          />
          <label
            class="Label__StyledLabel-sc-1pdmyea-0 fEfxuf"
            for="firstName"
            required=""
          >
            First Name
            <span
              class="Required__StyledRequired-sc-1du3aep-0 icgBVx"
            >
               *
            </span>
          </label>
          <input
            class="FieldText__Field-sc-1uqpfc5-0 dhVsEX"
            name="firstName"
            placeholder="Enter your first name"
            type="text"
            value=""
          />
          <label
            class="Label__StyledLabel-sc-1pdmyea-0 fEfxuf"
            for="lastName"
          >
            Last Name
          </label>
          <input
            class="FieldText__Field-sc-1uqpfc5-0 dhVsEX"
            name="lastName"
            placeholder="Enter your last name"
            type="text"
            value=""
          />
          <label
            class="Label__StyledLabel-sc-1pdmyea-0 fEfxuf"
            for="organization"
          >
            Organization Name
          </label>
          <input
            class="FieldText__Field-sc-1uqpfc5-0 dhVsEX"
            name="organization"
            placeholder="Enter your organization's name"
            type="text"
            value=""
          />
          <label
            class="Label__StyledLabel-sc-1pdmyea-0 fEfxuf"
            for="description"
            required=""
          >
            Description
            <span
              class="Required__StyledRequired-sc-1du3aep-0 icgBVx"
            >
               *
            </span>
          </label>
          <textarea
            class="FieldText__Field-sc-1uqpfc5-0 dhVsEX"
            name="description"
            placeholder="Enter a description for your slate"
            rows="8"
            type="textarea"
          />
        </div>
        <div
          class="create__Separator-sc-1iis3wx-0 izxoit"
        />
        <div
          class="pa4"
        >
          <div
            class="SectionLabel__StyledSectionLabel-sc-1fzxsyz-0 hMXPaX"
          >
            RECOMMENDATION
          </div>
          <label
            class="Label__StyledLabel-sc-1pdmyea-0 fEfxuf"
            for="recommendation"
            required=""
          >
            What type of recommendation would you like to make?
            <span
              class="Required__StyledRequired-sc-1du3aep-0 icgBVx"
            >
               *
            </span>
          </label>
          <div>
            <div
              class="Checkbox__Wrapper-sc-1ygebbu-0 guVRWe"
            >
              <div
                class="react-toggle"
              >
                <div
                  class="react-toggle-track"
                >
                  <div
                    class="react-toggle-track-check"
                  >
                    <svg
                      height="11"
                      viewBox="0 0 14 11"
                      width="14"
                    >
                      <title>
                        switch-check
                      </title>
                      <path
                        d="M11.264 0L5.26 6.004 2.103 2.847 0 4.95l5.26 5.26 8.108-8.107L11.264 0"
                        fill="#fff"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </div>
                  <div
                    class="react-toggle-track-x"
                  >
                    <svg
                      height="10"
                      viewBox="0 0 10 10"
                      width="10"
                    >
                      <title>
                        switch-x
                      </title>
                      <path
                        d="M9.9 2.12L7.78 0 4.95 2.828 2.12 0 0 2.12l2.83 2.83L0 7.776 2.123 9.9 4.95 7.07 7.78 9.9 9.9 7.776 7.072 4.95 9.9 2.12"
                        fill="#fff"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </div>
                </div>
                <div
                  class="react-toggle-thumb"
                />
                <input
                  class="react-toggle-screenreader-only"
                  name="recommendation"
                  type="checkbox"
                />
              </div>
              <label
                class="Checkbox__ToggleLabel-sc-1ygebbu-1 liSlWA Label__StyledLabel-sc-1pdmyea-0 fEfxuf"
                for="recommendation"
              >
                Recommend grant proposals
              </label>
            </div>
            <div
              class="Checkbox__Wrapper-sc-1ygebbu-0 guVRWe"
            >
              <div
                class="react-toggle"
              >
                <div
                  class="react-toggle-track"
                >
                  <div
                    class="react-toggle-track-check"
                  >
                    <svg
                      height="11"
                      viewBox="0 0 14 11"
                      width="14"
                    >
                      <title>
                        switch-check
                      </title>
                      <path
                        d="M11.264 0L5.26 6.004 2.103 2.847 0 4.95l5.26 5.26 8.108-8.107L11.264 0"
                        fill="#fff"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </div>
                  <div
                    class="react-toggle-track-x"
                  >
                    <svg
                      height="10"
                      viewBox="0 0 10 10"
                      width="10"
                    >
                      <title>
                        switch-x
                      </title>
                      <path
                        d="M9.9 2.12L7.78 0 4.95 2.828 2.12 0 0 2.12l2.83 2.83L0 7.776 2.123 9.9 4.95 7.07 7.78 9.9 9.9 7.776 7.072 4.95 9.9 2.12"
                        fill="#fff"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </div>
                </div>
                <div
                  class="react-toggle-thumb"
                />
                <input
                  class="react-toggle-screenreader-only"
                  name="recommendation"
                  type="checkbox"
                />
              </div>
              <label
                class="Checkbox__ToggleLabel-sc-1ygebbu-1 liSlWA Label__StyledLabel-sc-1pdmyea-0 fEfxuf"
                for="recommendation"
              >
                Recommend no action
              </label>
            </div>
          </div>
          <div>
            By recommending no action you are opposing any current or future slates for this batch.
          </div>
        </div>
        <div
          class="create__Separator-sc-1iis3wx-0 izxoit"
        />
        <div
          class="pa4"
        >
          <div
            class="SectionLabel__StyledSectionLabel-sc-1fzxsyz-0 hMXPaX"
          >
            GRANTS
          </div>
          <div
            class="mv3 f7 black-50"
          >
            Select the grants that you would like to add to your slate
          </div>
          <div
            class="flex flex-wrap"
          >
            <div
              class="Card__Wrapper-h3ymxi-0 gjicgu"
            >
              <div
                class="flex"
              >
                <div
                  class="Tag__Wrapper-l3n6sk-1 ceKKQx"
                >
                  <div
                    class="Tag__StyledTag-l3n6sk-0 cZzgaW"
                  >
                    GRANT PROPOSAL
                  </div>
                </div>
              </div>
              <div
                class="Card__CardTitle-h3ymxi-1 bSTxPd"
              >
                K Semantics for Web Assembly
              </div>
              <div
                class="Card__CardSubTitle-h3ymxi-2 kWUzrK"
              >
                200000
              </div>
              <div
                class="Card__CardDescription-h3ymxi-3 iIGibR"
              >
                A proposal to make Ethereum safer by enabling developers to formally verify smart contracts that run on Web Assembly.
              </div>
            </div>
            <div
              class="Card__Wrapper-h3ymxi-0 gjicgu"
            >
              <div
                class="flex"
              >
                <div
                  class="Tag__Wrapper-l3n6sk-1 ceKKQx"
                >
                  <div
                    class="Tag__StyledTag-l3n6sk-0 cZzgaW"
                  >
                    GRANT PROPOSAL
                  </div>
                </div>
              </div>
              <div
                class="Card__CardTitle-h3ymxi-1 bSTxPd"
              >
                Registry Builder Vote Weighting
              </div>
              <div
                class="Card__CardSubTitle-h3ymxi-2 kWUzrK"
              >
                175000
              </div>
              <div
                class="Card__CardDescription-h3ymxi-3 iIGibR"
              >
                A proposal to make Ethereum safer by improving a reusable library for token-curated registries.
              </div>
            </div>
            <div
              class="Card__Wrapper-h3ymxi-0 gjicgu"
            >
              <div
                class="flex"
              >
                <div
                  class="Tag__Wrapper-l3n6sk-1 ceKKQx"
                >
                  <div
                    class="Tag__StyledTag-l3n6sk-0 cZzgaW"
                  >
                    GRANT PROPOSAL
                  </div>
                </div>
              </div>
              <div
                class="Card__CardTitle-h3ymxi-1 bSTxPd"
              >
                Tidbit Dynamic Oracles
              </div>
              <div
                class="Card__CardSubTitle-h3ymxi-2 kWUzrK"
              >
                150000
              </div>
              <div
                class="Card__CardDescription-h3ymxi-3 iIGibR"
              >
                A proposal to make Ethereum safer by improving a reusable library for oracles.
              </div>
            </div>
            <div
              class="Card__Wrapper-h3ymxi-0 gjicgu"
            >
              <div
                class="flex"
              >
                <div
                  class="Tag__Wrapper-l3n6sk-1 ceKKQx"
                >
                  <div
                    class="Tag__StyledTag-l3n6sk-0 cZzgaW"
                  >
                    GRANT PROPOSAL
                  </div>
                </div>
              </div>
              <div
                class="Card__CardTitle-h3ymxi-1 bSTxPd"
              >
                Infrastructure Monitoring
              </div>
              <div
                class="Card__CardSubTitle-h3ymxi-2 kWUzrK"
              >
                200000
              </div>
              <div
                class="Card__CardDescription-h3ymxi-3 iIGibR"
              >
                A proposal to make Ethereum safer by monitoring high value systems and assets on Ethereum.
              </div>
            </div>
          </div>
        </div>
        <div
          class="create__Separator-sc-1iis3wx-0 izxoit"
        />
        <div
          class="flex pa4 justify-end"
        >
          <div
            class="Button__Wrapper-sc-1vedn6s-2 jHiEij"
          >
            <button
              class="Button__LargeButton-sc-1vedn6s-1 lfgnkA"
            >
              Back
            </button>
          </div>
          <div
            class="Button__Wrapper-sc-1vedn6s-2 jHiEij"
          >
            <button
              class="Button__LargeButton-sc-1vedn6s-1 lCuqq"
              type="submit"
            >
              Create Slate
            </button>
          </div>
        </div>
        <div />
      </form>
    </div>
  </div>
</div>
`;
